File: src/main/scala/hu/elte/inf/mbalassi/msc/spark/tricnt/DirectedTriangleCounter.scala
package hu.elte.inf.elte.mbalassi.msc.spark.tricnt;

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._

object DirectedTriangleCounter {
  def main(args: Array[String]) {
    
    val sparkHome = args(0)
    val input = args(1)
    
    val sc = new SparkContext("local", "DirectedTriangleCounter", sparkHome,
      List("target/scala-2.10/triangle-counter_2.10-1.0.jar"))

    val inputData = sc.textFile(input, 2)
    val graph = inputData.map(line => line.split(' ').map(vertex => vertex.toInt).toList)
                        .map(line => (line.head, line.tail)).cache

    val bidir = graph.map{case (vertex, links) => (links.filter(_ > vertex), vertex)}
    					.flatMap{case (targets, vertex) => targets zip List.fill(targets.length)(vertex)}
    val triCan = (graph.map{case (vertex, links) =>(vertex, links.filter(vertex > _))} join bidir)
                        .flatMap{case (vertex, (links, target)) => links zip List.fill(links.length)(target)}
    val triCnt = (graph join triCan).filter{case (vertex, (links, target)) => links.contains(target)}.count 
    println("TRI_CNT: %d".format(triCnt))
  }
}

File: src/main/java/hu/elte/inf/mbalassi/msc/seq/tricnt/DirectedTriangleCounter.java
package hu.elte.inf.mbalassi.msc.seq.tricnt;

public class DirectedTriangleCounter{

   private static long countDepthNInsts(int node, int goalNode, int depth,
                                        int[][] graph){
     if (depth == 0) return (node == goalNode ? 1 : 0);

     long cnt = 0;
     for (int dest = 0; dest < graph[node].length; ++dest){
         cnt += countDepthNInsts(graph[node][dest], goalNode, depth-1, graph);
     }
     return cnt;
   }
   
   public static long count(int[][] intGraph) {
       long triCnt = 0;

       int numNodes = intGraph.length;

       for(int node = 0; node < numNodes; ++node){
         triCnt += countDepthNInsts(node, node, 3, intGraph);
       }

       //every triangle is counted 3 times
       triCnt /= 3;
       
       return triCnt;
   }
}
